Zend/zend_object_handlers.c

```diff
@@ -805,7 +817,14 @@ ZEND_API zval *zend_std_write_property(zend_object *zobj, zend_string *name, zva
 
 	if (EXPECTED(IS_VALID_PROPERTY_OFFSET(property_offset))) {
 		variable_ptr = OBJ_PROP(zobj, property_offset);
+
 		if (Z_TYPE_P(variable_ptr) != IS_UNDEF) {
+			if (UNEXPECTED(Z_PROP_FLAG_P(variable_ptr) != IS_PROP_UNINIT && prop_info && prop_info->flags & ZEND_ACC_FINAL)) {
```

Comment:
I think this could just be `ZEND_ASSERT(prop_info)` before this (only for `IS_VALID_PROPERTY_OFFSET`). The prop_info should be set if a valid property reference was null, if I understand how this works correctly.

I'd guess it'd be more efficient as `UNEXPECTED(prop_info->flags & ZEND_ACC_FINAL && Z_PROP_FLAG_P(variable_ptr) != IS_PROP_UNINIT)`, because ACC_FINAL would probably be much less common than being initialized.

```
#define IS_VALID_PROPERTY_OFFSET(offset)           ((intptr_t)(offset) > 0)
#define IS_WRONG_PROPERTY_OFFSET(offset)           ((intptr_t)(offset) == 0)
#define IS_DYNAMIC_PROPERTY_OFFSET(offset)         ((intptr_t)(offset) < 0)
```

Reply:
I'll change the order of the condition, thank you for the pointing this out!

However, it seems to me that it's not possible to add a `ZEND_ASSERT(prop_info)` because as far as I saw, it's only set for typed properties. And the `if (UNEXPECTED(prop_info))` condition some lines below also confirm this I think.